File: Unity/AI_FSM/Assets/Script/Transition/WaitTransition.cs
using System.Collections;
using System.Collections.Generic;
using System.Timers;
using UnityEngine;

[CreateAssetMenu(fileName = "WaitTransition", menuName ="FSM/Transition/Create WaitTransition")]
public class WaitTransition : Transition
{
	[SerializeField, Range(0, 5)] float waitTime = 3;

	bool isDone = false;

	public override bool IsValidTransition => isDone;

	public override void InitTransition(FSM _owner)
	{
		base.InitTransition(_owner);
		_owner.Owner.StartCoroutine(Wait());
	}
	IEnumerator Wait()
	{
		yield return new WaitForSeconds(waitTime);
		isDone = true;
	}
}

File: Unreal/IA_FSM/Source/IA_FSM/Private/FSM/Transition/WaitTransition.cpp
// Fill out your copyright notice in the Description page of Project Settings.


#include "FSM/Transition/WaitTransition.h"

void UWaitTransition::InitTransition()
{
    GetWorld()->GetTimerManager().SetTimer(waitTimer, this, &UWaitTransition::Wait, waitTime, false);
}

bool UWaitTransition::IsValidTransition()
{
    return isDone;
}

void UWaitTransition::Wait()
{
    isDone = true;
}
